src/arrays.jl

```diff
@@ -31,7 +32,7 @@ struct QuantityArray{T,N,D<:AbstractDimensions,Q<:AbstractUnionQuantity{T,D},V<:
     dimensions::D
 
     function QuantityArray(v::_V, d::_D, ::Type{_Q}) where {_T,_N,_D<:AbstractDimensions,_Q<:AbstractUnionQuantity,_V<:AbstractArray{_T,_N}}
-        Q_out = constructor_of(_Q){_T,_D}
+        Q_out = constructorof(_Q){_T,_D}
```

Comment:
IMO this is wrong in general (both the old and new version). There's no guarantee that `consteuctorof` returns a parameterless type and even if in general  these type parameters might be wronf or not even exist.

Reply:
Not sure if this applies here since it is guaranteed that `Q <: AbstractUnionQuantity`? Also `T` and `D` are also required by the function.

Reply:
Maybe I should just rename it to be `base_quantity` or something, and get people to overload it? (Rather than `constructor`, which maybe is not appropriate)

Reply:
IMO these methods should just not be defined in general but only for concrete types such as `Quantity` - and in that case I think it would be preferable and easier to just call the desired constructor (`Quantity`, `Quantity{T}` or `Quantity{T,D}`) directly.

For other subtypes these type parameters might not even exist - a subtype of `AbstractQuantity{T,D}` can hardcode `T` and `D` to specific types. Moreover, even if `T` and `D` are still type parameters it is not guaranteed that `T` and `D` are also the first and second type parameter of the subtype.

Reply:
I see what you mean but I think this might just be too complicated to be worth it. If the user wants to make an `AbstractQuantity{T,D}` and  not have `T` be the value type, then they would break several assumptions throughout the `QuantityArray` interface (which sets `T` based on the array you pass in).

Idk. I guess I'd rather have things break quickly and in an interpretable way via failing to build this constructor, rather than try to accommodate some very complex behavior for some user-defined struct that does not even exist.

So my vote is to just leave this as a TODO for later. Sound good?

Reply:
My suggestion would be to directly fix these issues: You just have to restrict this function to `_Q <: Union{Quantity, ...}` (whatever other concrete types exist in DynamicQuantities).

Reply:
Do you mean: disallow any abstract quantity that is not `Quantity` or `GenericQuantity`?

Or, do you mean a trait-based approach like writing a custom function like

```julia
with_param_types(::Type{<:AbstractUnionQuantity}, _, _) = error("Please write a custom method")
with_param_types(::Type{<:Quantity}, ::Type{T}, ::Type{D}) = Quantity{T,D}
with_param_types(::Type{<:GenericQuantity}, ::Type{T}, ::Type{D}) = GenericQuantity{T,D}
```

This seems clean. Wdyt?

Reply:
I would only support `Quantity` and `GenericQuantity`. My general view is that code should be as simple as possible and traits and other generalizations should only be added if there is a clear use case for them.

Reply:
Sg, added.